File: python/medium/design_underground_system.py
# https://leetcode.com/problems/design-underground-system/

class UndergroundSystem:

    def __init__(self):
        self.journey = {}
        self.history = {}
        

    def checkIn(self, id: int, startStation: str, startTime: int) -> None:
        self.journey[id] = (startStation, startTime)
        

    def checkOut(self, id: int, endStation: str, endTime: int) -> None:
        startStation, startTime = self.journey.pop(id)
        key = (startStation, endStation)
        allTime, allCount = self.history.get(key, (0, 0))
        self.history[key] = (allTime + (endTime - startTime), allCount + 1)
        

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        key = (startStation, endStation)
        allTime, allCount = self.history.get(key, (0, 0))
        return allTime / allCount
        

# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)

File: rust/medium/design_underground_system.rs
// https://leetcode.com/problems/design-underground-system/ 

use std::collections::HashMap;

#[derive(Default)]
struct UndergroundSystem {
    journey: HashMap<i32, (String, i32)>,
    history: HashMap<(String, String), (i32, u32)>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl UndergroundSystem {

    fn new() -> Self {
        Default::default()
    }
    
    fn check_in(&mut self, id: i32, start_station: String, start_time: i32) {
        self.journey.insert(id, (start_station, start_time));
    }
    
    fn check_out(&mut self, id: i32, end_station: String, end_time: i32) {
        if let Some((start_station, start_time)) = self.journey.get(&id) {
            let travel = self
                .history
                .entry((start_station.clone(), end_station))
                .or_default();
            travel.0 += end_time - start_time;
            travel.1 += 1
        }
    }
    
    fn get_average_time(&self, start_station: String, end_station: String) -> f64 {
        if let Some(&(total_time, count)) = self.history.get(&(start_station, end_station)) {
            return f64::from(total_time) / count as f64
        }
        unreachable!()
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * let obj = UndergroundSystem::new();
 * obj.check_in(id, stationName, t);
 * obj.check_out(id, stationName, t);
 * let ret_3: f64 = obj.get_average_time(startStation, endStation);
 */
